### PR TITLE
New expando parser

### PR DESCRIPTION
**Pull request by g0mb4**
New expando parser based on #3937

**Nodes:**
- [x] text
- [x] expando
- [x] pad
- [x] condition

**Formats:**
- [x] alias_format_str()
- [x] attach_format_str()
- [x] autocrypt_format_str()
- [x] compose_format_str()
- [x] compress_format_str()
- [x] crypt_format_str()
- [x] folder_format_str()
- [x] greeting_format_str()
- [x] group_index_format_str()
- [x] history_format_str()
- [x] index_format_str()
- [x] mix_format_str()
- [x] nntp_format_str()
- [x] pattern_format_str()
- [x] pgp_command_format_str()
- [x] pgp_entry_format_str()
- [x] query_format_str()
- [x] sidebar_format_str()
- [x] smime_command_format_str()
- [x] status_format_str()

**Render:**
- [x] Get one simple expando working
- [x] Get all expandos working
- [x] Fix padding error
 
**Tests:**
- [x] Tests for each config type
- [x] Tests of formatting
- [x] Test for 0, '\0', and ' ' in `if` render

**Misc:**
- [x] custom config type
- [x] tests
- [x] TODOs / FIXMEs
- [ ] docs

**TODO:**
- [x] Fix formatted dates
- [x] Fix formatted ```index-fomat-hook```s
- [x] Test utf-8 padding
- [ ] Refactor
- [x] separate `MUTT_FLAG_INDEX` from color in softpadding
- [ ]  Refactor ```$arrow_cursor```

### CODE REVIEW

**Comment by flatcap** [comment]
Rebased over `main`

**Comment by flatcap** [comment]
In `neomutt_free()` you need to change:
```diff
-  FREE(n->expando_table);
+  FREE(&n->expando_table);
```

This leak is causing all the tests to fail under ASAN.

**Comment by flatcap** [comment]
The `neomutt_free()` test was failing because it was bypassing `neomutt_new()`.
The test is there for coverage more that actual testing.

**Comment by g0mb4** [comment]
Thanks! You saved a lot of time for me :)

**Comment by flatcap** [comment]
- Rebased over `main`
- Moved all the `print_*` functions into `debug.c`
- Reverted the `.vscode` addition to `.gitignore` -- that should live in `~/.gitignore`
- Reverted some `clang-format`ing of headers -- we only do the source code
  Adding extra whitespace to headers, to align function names, etc, makes it easier for humans to read.

I've squashed down the PR to just four commits -- otherwise my brain will explode :-)

- 7841c223b build: reduce number of builds
- d41ef78fd wip expando parser
- 02b5a8cce unit tests
- a7605edd7 test main

I've archived your entire history to the [neomutt-old](https://github.com/neomutt/neomutt-old) repo, branch [expando-parser](https://github.com/neomutt/neomutt-old/tree/expando-parser).

**Comment by flatcap** [comment]
I've had a read through and I'm starting to follow it all...

It's time to start splitting up everything and I think that means reworking `ExpandNode`.
Because this is C and we can't do inheritance, we've already got some duplication, e.g.

https://github.com/neomutt/neomutt/blob/a7605edd751ddcddb1c07f024446cc8351191815/expando/parser.h#L19-L34

If we want to split up the parsing, this means adding more functions, e.g. `new` and `free` to `ExpandNode`, which means more duplication.

The workaround which we use elsewhere in NeoMutt is to have **one** `struct` which has "private data":

```c
void *ndata;
void (*ndata_free)(void **ptr);
```

See, [`struct Mailbox`](https://github.com/neomutt/neomutt/blob/main/core/mailbox.h#L78-L145), [`struct MuttWindow`](https://github.com/neomutt/neomutt/blob/main/gui/mutt_window.h#L122-L182)

**Comment by g0mb4** [comment]
Thanks! I'll start the refactoring in a couple if days.

**Comment by ossilator** [comment]
> Because this is C and we can't do inheritance,

oh, but we totally _can_ ... at the cost of some ugliness.
see https://sourceforge.net/p/isync/isync/ci/c3d91ae1e87f8a8b8d6b0e7ecd4b7fc09837d30e for example; either side of the patch shows a workable approach.

fwiw, given that isync is somewhat object-oriented, i've been pondering just using c++. the migration would happen in steps: first proper inheritance, then converting functions to methods and callbacks to virtuals, and at some point possibly stl containers. not sure it's worth it.

**Comment by ossilator** [comment]
> > i've been pondering just using c++
> 
> That's a future step for neomutt, too.

if you're confident about it, then i'd recommend to just do the switch immediately, and start using c++ for new and refactored code. bulk conversions can be done later.

**Comment by lucilanga** [comment]
> i'd recommend to just do the switch immediately, and start using c++ for new and refactored code. bulk conversions can be done later.

Please! 
do enlighten us, why would you recommend something like that for this kind of project and for the state that is currently in ?

**Comment by ossilator** [comment]
if the switch is planned anyway, then doing it sooner rather than later saves churn.

**Comment by lucilanga** [comment]
Ultimately the gunman here is right!
Is the switch planned anyway ?
If so ... well it might not be everyone's line of thought and obviously not something you can pop over the night. I know this has been brought before, but still a valid question.

**Comment by flatcap** [comment]
Sorry, C++ is not going to happen in the near future.
There's still too much wrong in the code that needs fixing.
e.g. more APIs between sections of code to isolate them.

Yes, I have a handful of new devs who are skilled and motivated -- I'm really grateful for them.
I don't mind whether they work on refactoring or adding small features, because they're learning the code.
However, each area they work on requires **me** to understand that code too, and the context switches are time-consuming.

If you'd like to discuss C++ further, or have ideas how to improve NeoMutt's architecture, please open a [Discussion](https://github.com/neomutt/neomutt/discussions/categories/development).

**Comment by ossilator** [comment]
> There's still too much wrong in the code that needs fixing.

that's undoubtedly true, but i don't see how it gets in the way - quite the opposite. for me, using minimal c++ would be just a cleaner way to express what i want to do anyway, and i sort of get the vibe that you think the same.

i'm not saying that your plan is wrong, but i think it would make sense to double-check the reasoning and spell out the assumptions behind it. of course not here, because it's obviously a tangent.

**Comment by flatcap** [comment]
> i don't see how it gets in the way

Firstly, there aren't many independent parts of the code.
If you try to change parts of, say IMAP, then the repercussions will affect half the codebase.

Second, I don't want to spend time mediating discussions about _this_ C++ feature or _that_.
When there's someone, besides me, who has a handle on the 150,000 lines of spaghetti, maybe then.

If anyone has concrete ideas how it can be done, without putting too much burden on **me**, please open a discussion.

Now, please can we give the PR back to g0mb4.

**Comment by g0mb4** [comment]
Dear @flatcap ! 

The current version is capable of displaying the default `$index_format` correctly as far as I can see.  I'll start to implement an other format string (`$status_format`) to see what patterns/functions will be common.  Can you look at the redesigned `ExpandoNode` please? The formatting callbacks are a mess at the moment, but I'd like to finalize the data structures first. Thank you!

**Comment by flatcap** [comment]
Thanks!  I'm still working through the code.

Meanwhile, the conditional in `$index_format` (`%<l?%4l&%4c>`) seems to have stopped working in the last couple of commits.

```diff
-2     Oct 11 apple           (0.9K) apple apple APPLE
+2     Oct 11 apple           () apple apple APPLE
```

**Comment by flatcap** [comment]
I've done a little tidying...

- Rebased over `main`

- 0d69edba5 tidy large enum, table
  - Sort `enum ExpandoFormatIndex`
  - Sort and Format `expando_validation[]`
  Generally, we want to let `clang-format` organise the code, but in the case of large tables sometimes we make exceptions.
 
- 7d3670722 tidy code
  - Kill trailing whitespace
  - include "config.h" in `*.c`
  - Create library header
    Expose everything via a single `expando/lib.h`
  - Use library includes
    Include **library** headers, e.g. `mutt/lib.h` rather than individual files
  - Run [iwyu](https://neomutt.org/dev/analysis/iwyu)

- 0a225b888 doxy: link up source/library
  Eventually, everything will be documented.
  Add the bare bones of the library docs.

**Comment by g0mb4** [comment]
Wow! Thank you!

**Comment by g0mb4** [comment]
Same :D 

BTW, can you help me please? I think ade46c5ae07f862101d233aa813131feff31935b broke the tests in this branch and I'm not familiar with the CI.

**Comment by g0mb4** [comment]
Thanks!

**Comment by g0mb4** [comment]
Dear @flatcap !

I'm slowly working throug the refactoring and I'm arrived to [history_format_str()](https://github.com/neomutt/neomutt/blob/a2917814bcae00cf294c2596d82f00282e60d7cc/history/dlg_history.c#L85-L91). Do we want to keep it as an expando? We can use a simple `mutt_format_s` in that case.

**Comment by flatcap** [comment]
It may seem ridiculous, but I think the answer's, "Yes".
That way, History's `Menu` works the same as all the other `Menu`s.
`Menu.make_entry()` -> `history_make_entry()` -> format.

**Comment by g0mb4** [comment]
Ok, I see your point.

**Comment by g0mb4** [comment]
Dear @flatcap !

I'm almost ready with the new parser (just have to figure out how to save the parsed hooks and do the documentation). I think is is functionally complete now.  Would you be so kind to take a look at it?

**Comment by flatcap** [comment]
Rebased over `main`, tidied and squashed
The full 196-commit history is in [neomutt-old](https://github.com/neomutt/neomutt-old).

A couple of things caused me trouble (only briefly):

- Blank lines in the middle of blocks of `#include`s
  (clang-format isn't smart enough to handle _that_)

- Lots of files that didn't end in `\n`
  You may want to check you editor settings.

**Comment by flatcap** [comment]
So... I think the next step is to break up the mega-dependencies of `expando/validation.c`

I suggest and `expando.c` for each `config.c`.
It can contain all the data defining each expando.

Then, we can put a `struct ExpandoValidation` in the `struct ConfigDef.data` field (see `$mbox_type`, `MboxTypeDef`).

After that, we need to work out how to factor out the custom parsing, e.g. 'index-format-hook`

File: docs/config.c [hunk 1]
```diff
@@ -1710,6 +1710,17 @@
...
 */
 
+{ "history_format", DT_STRING, "%s" },
+/*
+** .pp
+** Controls the format of the entries of the history list.
+** This string is similar to $$index_format, but has its own
+** set of \fCprintf(3)\fP-like sequences:
+** .dl
+** .dt %s .dd History match
```
Comment (flatcap) [inline docs/config.c:1720]: Oh!  Good idea.
How about adding `%C` for the row number?
Jumping, by typing `123<enter>` is already supported.

**Comment by g0mb4** [comment]
Hmm, so you want `libexpando` to depend on nothing. I'll try to reduce the number of dependencies.

File: docs/config.c (hunk 1 above)
Comment (g0mb4) [reply docs/config.c:1720]: I had to store the `ExpandoRecord` somewhere :D Sure, I'll add that.

Comment (g0mb4) [reply docs/config.c:1720]: I'm a bit lost, I totally forgot how to open the history dialog, can you help me please?

**Comment by flatcap** [comment]
> so you want `libexpando` to depend on nothing

Pretty much, yes.  Just `libmutt` and maybe `libconfig`

`libexpando` will contain the code to parse the basic expandos, validate the string, build the tree
and generate a formatted string (using all your callback functions).

All the data that defines a particular format string will be external.

Some format strings (index-format-hook) will need a custom **parser**.
That too will need to be outside of `libexpando` (still thinking how).

**Comment by g0mb4** [comment]
Thanks! I'm a bit busy right now, but I'll continue the development in about a week.

**Comment by g0mb4** [comment]
Dear @flatcap !

I came up with a solution to make libexpando independent. Can you check it out please?
Only `$index_format` is implemented just to see if it is a good direction. I'll fix the tests later if this solution stays.

**Comment by flatcap** [comment]
Thanks.  It looks a bit better, but there's still some global-ness about it.
I had a think and started to make some notes.
I'll try to write them up later (after my crossword :-)

**Comment by flatcap** [comment]
I've created a full work-through -- sorry for not going it sooner.
I hope it makes sense.
As always, ask lots of questions :-)

**Comment by flatcap** [comment]
I haven't heard from you in a while -- I hope you're well.
Hopefully, you've been busy with "real life", rather than being burnt-out :-)

I've added a couple of commits that define **every** expando,
and link the 51 config variables to the 18 format functions.
I've added [notes to the discussion](https://github.com/neomutt/neomutt/discussions/3937#discussioncomment-7532833).

You've done an amazing job so far, and I think we've reached the high-point of complexity.
From here on, I think everything will start getting smaller and simpler.

**Comment by g0mb4** [comment]
Dear @flatcap ! Unfortunately I'm a bit overwhelmed in work. I did not forget about the project, but I need some time to sort things out.

**Comment by flatcap** [comment]
> I'm a bit overwhelmed in work

No worries

**Comment by flatcap** [comment]
OK, I've been busy... :-)
I've refactored the commits into some sensible steps.
(see the commit messages for more details)

Inside the library, I haven't changed much.
A little light reorganising and I renamed `ExpandoRecord` to just `Expando`.

The config type has had a bit of an overhaul to correct a misunderstanding about validators.
The **set** methods do the parsing and if they fail, they log to `Buffer *err`.
This leaves the validator unset for most config variables.
The validator could be used for business logic, e.g. String must contain `%s`

Most of my changes are outside the library.

I defined data **domains** (23).  Each one correlates to `struct` in the code.
e.g. `ED_ALIAS`

Each domain has **UID**s that identify a single member (about 200 in all).
e.g. `ED_ALI_COMMENT`

Next, I defined `ExpandoDefinition`s for each domain.
These are linked in the data field for each config variable.
e.g. `AliasFormatData`
```c
{ "c", "comment", ED_ALIAS, ED_ALI_COMMENT, E_TYPE_STRING, E_FLAGS_NO_FLAGS, NULL },
```

Finally, I defined `ExpandoRenderData`s for each caller of `expando_render()`.
These link the Domain-UID pairs to a callback function.
e.g. `AliasRenderData`
```c
{ ED_ALIAS, ED_ALI_COMMENT, alias_c },
```

For the majority of the config variables, I _think_ everything's in place, now.
The library should have access to all the info it needs to render the strings.

Ping me when you have some free time.
Thanks!

**Comment by flatcap** [comment]
Happy New Year, @g0mb4 

Has work calmed down yet?
I'd love to get this working and merged!

If you have time, drop in to IRC and we can discuss what's next and how I can help.
Thanks!

**Comment by g0mb4** [comment]
Dear @flatcap!

Happy new year to you as well! I think I'll have some free time to finish this PR. I'll read your changes and redesign the architecture. It will take a couple of days. In the meantime can you please write a short summary of your expectations moving forward?

**Comment by flatcap** [comment]
The actual expando code is still a black box to me, but...

Before my adjustments, `$index_format` "worked".
There was enough data in its `ConfigDef` to render a string.

I've reorganised the data in `ConfigDef`, so the expando code needs updating,
but there's now data for **every** Expando config option. ([see notes above](https://github.com/neomutt/neomutt/pull/3944#issuecomment-1813684925))
Hopefully, it'll be a quick job to get it running again.

## What next?

Some of the recent non-expando code changes have made rebasing tricky,
so I'd like to get this merged as soon as we can.

Tickbox time :-)

- [ ] Get one simple expando working
- [ ] Get all expandos working (so NeoMutt _looks_ OK when run)
- [ ] Fix tests
  - Tests for each config type
  - Tests of formatting

## Releasing the new Expandos

We're due a bug-fix release, very soon.
After that, if the Expandos are looking good,
I _might_ make a release with **just** the Expando code.

---

As always, ask lots of questions (here/irc)

Thanks :heart:

**Comment by g0mb4** [comment]
Dear @flatcap!

I've implemented the rendering of single expandos and a custom parser for dates. Can you please comment on the new `*render.c` test files?

**Comment by flatcap** [comment]
Rebased over more config changes.

> Can you please comment on the new *render.c test files?

On it.

**Comment by flatcap** [comment]
They look really good.
I've just tweaked the date test to  be a number other than 0 to make it obvious that it's working.
(2016-03-07 is NeoMutt's first release).

I've also added a line to `Makefile.autosetup` to enable code coverage.
See: https://neomutt.org/dev/analysis/coverage
It's often useful to see which bit of the code are, or aren't, being used by tests.
(Don't obsess over the numbers, though :-)

**Comment by g0mb4** [comment]
Dear @flatcap!

I think I've implemented the new parser + renderer. 

Some notes:
+ I did not use  `ExpandoDataType` and  `ExpandoFlags` (maybe these fields can be removed)
+ I think all of the expandos are working, but I'm not sure. 

Next tasks:
+ I'll implement more rendering tests. 
+ I'll document the code.

Please review the code.

**Comment by flatcap** [comment]
My `index_format` doesn't work.
It's quite complicated, containing nested dates.

> hdrline.c:1792: mutt_make_string: Assertion `exp' failed

`neomutt -n -F test.rc`

**test.rc**
```
set index_format = "%<[12m?%<[7d?%<[12H?%[%H:%M ]&%[%a-%d]>&%[%d-%b]>&%[%b-%y]> %Z %-25.25F %<M?+%M) & >%s %* %<M?&(%c)>"
```

**Comment by flatcap** [comment]
A simpler case is just the last bit: `%<M?&(%c)>"`
If `%M` (thread is collapsed) display **nothing**.
Otherwise display `(%c)` (bytes)

Change it to `%<M?X&(%c)>"` and it works.

**Comment by g0mb4** [comment]
Ohh, I see the problem. My code assumes there is always a valid non-empty
expression in the true-case. I'll fix it tomorrow.

Richard Russon ***@***.***> ezt írta (időpont: 2024. jan.
23., K 17:36):

**Comment by flatcap** [comment]
🔴 Functions that are called from outside of `libexpando`.



Source: [gv](https://github.com/neomutt/test-gfx/blob/main/2024-01-25_expando/expando-helpers.gv) [svg](https://github.com/neomutt/test-gfx/blob/main/2024-01-25_expando/expando-helpers.svg)



Source: [gv](https://github.com/neomutt/test-gfx/blob/main/2024-01-25_expando/expando-config.gv) [svg](https://github.com/neomutt/test-gfx/blob/main/2024-01-25_expando/expando-config.svg)



Source: [gv](https://github.com/neomutt/test-gfx/blob/main/2024-01-25_expando/expando-render.gv) [svg](https://github.com/neomutt/test-gfx/blob/main/2024-01-25_expando/expando-render.svg)



Source: [gv](https://github.com/neomutt/test-gfx/blob/main/2024-01-25_expando/expando-parse.gv) [svg](https://github.com/neomutt/test-gfx/blob/main/2024-01-25_expando/expando-parse.svg)

**Comment by flatcap** [comment]
Let me know if you need any feedback

**Comment by flatcap** [comment]
OK, I've found the problem I'm having.
Not sure of the fix, yet.

In `expando/format_callbacks.c`, the callback is passing `MUTT_FORMAT_NO_FLAGS`:

https://github.com/neomutt/neomutt/blob/372e854a3a4d901879f8c4f4457b0cc12b085667/expando/format_callbacks.c#L176-L177

In `index_M()` we're checking:

https://github.com/neomutt/neomutt/blob/372e854a3a4d901879f8c4f4457b0cc12b085667/hdrline.c#L1117

which fails, so we don't get the true case:

https://github.com/neomutt/neomutt/blob/372e854a3a4d901879f8c4f4457b0cc12b085667/hdrline.c#L1121-L1126

**Comment by flatcap** [comment]
this seems to do the trick:
```diff
-      printed = rd->callback(cp->condition, tmp, sizeof(tmp), sizeof(tmp), data, MUTT_FORMAT_NO_FLAGS);
+      printed = rd->callback(cp->condition, tmp, sizeof(tmp), sizeof(tmp), data, flags);
```

Now my index looks perfect!

**Comment by g0mb4** [comment]
Really nice! This was the next thing I wanted to try :D It also proves that some config specific test are needed, at least for `$index_format`.

**Comment by flatcap** [comment]
I've done a bit of tidying, rebased and squashed.
The main changes are:

- Rename `struct ExpandoNode` `self` to `node` for clarity
- Change private data from `intptr_t` to `void *`
  `intptr_t` was my original choice, but the data will always be a pointer.
- Remove lots of casts
  Using `void *` means we can drop **so many** casts
- Change `format_*()` helpers to take `ExpandoNode`
  This reduces the amount of `Expando'` knowledge the callback needs.

**Comment by g0mb4** [comment]
Dear @flatcap !

I think it is done. It needs more testing, I'm sure I've missed some special cases...

**Comment by g0mb4** [comment]
Dear @flatcap !

I'm back. What would be the next step?

**Comment by flatcap** [comment]
Great!
I've been busy thinking and I've got pages of notes! :-)
If you've got a sec to chat, so I don't overwhelm you.
Thanks

**Comment by flatcap** [comment]
I was going to write them all up, here, anyway.
I just wanted a chance to discuss the ideas first.

**Comment by g0mb4** [comment]
Oh, okay. Unfortunately I won't be able to log in to chat today, but I'll be available tomorrow, if it is okay.

**Comment by flatcap** [comment]
Not yet implemented: `index-format-hook`
Sample config:
```sh
index-format-hook  date  "~d<1d"    "AAA %[%H:%M]"
index-format-hook  date  "~d<1m"    "BBB %[%a %d]"
index-format-hook  date  "~d<1y"    "CCC %[%b %d]"
index-format-hook  date  "~A"       "DDD %[%m/%y]"

set index_format="%4C %-6@date@ %-15.15F %Z (%4c) %s"
```

**Comment by flatcap** [comment]
## Expando Types

Empty, Text, Expando, Padding, Conditional, Conditional Date, Container.

Strip down `ExpandoNode` to:

```c
struct ExpandoNode
{
  enum ExpandoNodeType type;
  struct ExpandoNode  *next;

  void *ndata;
  void (*ndata_free)(void **ptr);
};
```

Each node type has its own ndata.

### Empty

No private data :-)

### Text

```c
struct ExpandoNodeText
{
  const char *start;
  const char *end;
};
```

### Padding

```c
struct ExpandoNodePad
{
  enum ExpandoPadType pad_type;
  char                pad_char;
}:
```

### Conditional

```c
struct ExpandoNodeConditional
{
  struct ExpandoNode *condition;
  struct ExpandoNode *if_true_tree;
  struct ExpandoNode *if_false_tree;
};
```

### Conditional Date

```c
struct ExpandoNodeConditionalDate
{
  int    count;
  char   period;
  time_t multiplier;
};
```

### Container

```c
struct ExpandoNodeContainer
{
  struct ExpandoNode *children;
};
```

**Comment by flatcap** [comment]
## Custom Nodes

If an `struct ExpandoDefinition` has a custom parser, it may return a custom `ExpandoNode`, with a custom child.

### Custom

```c
struct ExpandoNode
{
  ENT_CUSTOM;
};
```

```c
struct ExpandoNodeMyType
{
  const char *data1;
  int         value2;
};
```

**Comment by flatcap** [comment]
## Expando Node

Upgrade `ExpandoNode` to have an API.

We want to be able to render the string, test for "true-ness" and possibly generate a new format string.

```c
struct ExpandoNode
{
  enum ExpandoNodeType type;
  struct ExpandoNode  *next;

  void *ndata;
  void (*ndata_free)(void **ptr);

  /**
   * render - Render an Expando
   * @param[in]  node      ExpandoNode to format
   * @param[in]  max_width Number of screen columns
   * @param[in]  flags     Flags, see #MuttFormatFlags
   * @param[out] buf       Buffer for the result
   * @retval num Number of bytes written to buf
   *
   * Each callback function implements some expandos, e.g.
   *
   * | Expando | Description
   * | :------ | :----------
   * | \%t     | Title
   */
  int (*render)(const struct ExpandoNode *node, int max_width, MuttFormatFlags flags, struct Buffer *buf);

  /**
   * is_true - Test a conditional Expando
   * @param[in]  node      ExpandoNode to test
   * @retval true Expando condition is true
   */
  bool (*is_true)(const struct ExpandoNode *node);

  /**
   * generate - Create a new format string
   * @param[in]  node ExpandoNode to formay
   * @param[out] buf  Buffer for the result
   * @retval num Number of bytes written to buf
   */
  int (*generate)(const struct ExpandoNode *node, struct Buffer *buf);
};
```

- `render()`
  Allow the expando code to render nodes without knowing what type they are.

- `is_true()`
  Many expandos can be used conditionally.
  e.g. Numbers that are greater than zero; Strings that are non-empty
  A dedicated test function allows us to delegate the tricky cases, like conditional dates.

- `generate()` - In the future
  Generate a format string from the Node tree.
  Many programs allow you to tweak the order and size of table columns.

**Comment by flatcap** [comment]
## Colouring

Currently, the colours are magic control codes that are inserted into the text.
The Expando code needs to recognise them and ignore them when calculating widths.

Only Expando Nodes are coloured and only `$index_format`s `%s` requires two colours.
If `%s` is split into two pieces, then colouring can be done more easily.

If the render function can return three strings, the expando code can make smarter decisions.

- Render `%X` -\> `[colour]` + `John Smith` + `[end-colour]`

The prefix/suffix parts can be considered zero-width.
The text can be truncated if it needs to be, leading to:

- `[colour]` + `John Sm` + `[end-colour]`

**Comment by flatcap** [comment]
## Resizing Nodes

There are over 200 expando callbacks.
Most of them need to do some form of formatting.

The render function knows how much space is available, so it makes sense for it to do the formatting.

### Simplify the callbacks

```c
int index_L(const struct ExpandoNode *node, char *buf, int buf_len, int cols_len, void *data, MuttFormatFlags flags)
{
  assert(node->type == ENT_EXPANDO);

  const struct HdrFormatInfo *hfi = data;
  const struct Email *e = hfi->email;

  char fmt[128] = { 0 };
  char tmp[128] = { 0 };

  make_from(e->env, tmp, sizeof(tmp), true, flags);
  format_string(fmt, sizeof(fmt), tmp, flags, MT_COLOR_INDEX_AUTHOR,
                MT_COLOR_INDEX, node, NO_TREE);
  return snprintf(buf, buf_len, "%s", fmt);
}
```

might become:

```c
int index_L(const struct ExpandoNode *node, MuttFormatFlags flags, struct Buffer *buf)
{
  assert(node->type == ENT_EXPANDO);

  const struct HdrFormatInfo *hfi = node->ndata;
  const struct Email *e = hfi->email;

  return make_from(e->env, buf, true, flags);
}
```

### Resizing all the Nodes

Many Windows in NeoMutt are dynamically sized.
They can be MINIMISE, FIXED or MAXIMISE.
Each requests an amount of space and the reflow functions allocate it fairly.

The Expandos could be a simpler version of the Windows.

The first pass would be to collect all the strings into Nodes.

- `%a%b%* %c` -> `apple` + `banana` + Pad-space + `cherry`

We have some Padding, so we'll need to check the available space.
There's enough room for `%a`, but not all of `%b`, so we get:

- `apple` + `bana` + Pad-space + `cherry`

which renders to:

- `applebanacherry`

With colours that might be:

- `[+a]apple[-a][+b]bana[-b][+c]cherry[-c]`

**Comment by g0mb4** [comment]
Dear @flatcap!

I've finished fixing some UTF-8 related bugs and I think I'm ready to start thinking about the refactoring.
I've added ```ExpandoNotes.md```, we can use that file as a reference so it would *always* contain the most recent version of our ideas. (Searching through the comments can be time consuming).

**Comment by flatcap** [comment]
> I think I'm ready to start thinking about the refactoring

Great!

> I've added `ExpandoNotes.md`, ... the most recent version of our ideas.

Good idea.
I'll have a read in a second.

> (Searching through the comments can be time consuming).

Yeah :-)

**Comment by flatcap** [comment]
I've pushed a bunch of `Buffer` changes under yours.
I'm _mostly_ happy with them, and your code still works :-)
I'll merge them soon.

I've fixed a few leaks in the parser.
Coverity's happy, but you may wish to review them.

I've fixed the behaviour of a couple of browser date handlers.

Finally, I've split out a change you made to `test/Makefile.autosetup`.
BSD isn't happy with it.

**Comment by flatcap** [comment]
Quick hack: 0c18ce62a Graphviz.
Hitting <kbd>p</kbd> (`<print-message>`) will dump `$index_format` to a timestamped file.



Source: [gv](https://github.com/neomutt/test-gfx/blob/main/2024-01-25_expando/14%3A07%3A31-expando.gv) [svg](https://github.com/neomutt/test-gfx/blob/main/2024-01-25_expando/14%3A07%3A31-expando.svg)

**Comment by g0mb4** [comment]
Wow! I really like the graph!

**Comment by flatcap** [comment]
the last two commits do a bit of vim-fu magic to fix the prototypes of all the callbacks, to match your change:

```diff
-typedef int (*expando_render_t)(const struct ExpandoNode *node, char *buf, int buf_len, int cols_len, void *data, MuttFormatFlags flags);
+typedef void (*expando_render_t)(const struct ExpandoNode *node, void *data, MuttFormatFlags flags, struct Buffer *buf);
```

**Comment by flatcap** [comment]
- fix up the build and tests
- sort out the callback naming / comments
- changed `expando_render()` to take a `Buffer`
- unify the prototype of the render functions
- created a temp build fix for `format_*()`

**Comment by flatcap** [comment]
A little tidying and a LOT of renaming / reorganising.
I've split the Node types into separate files and named everything accordingly.
The parser and render functions still need a little work to simplify them.

**Comment by g0mb4** [comment]
formats using ```$arrow_cursor```:

- alias_format
- attach_format,
- autocrypt_acct_format
- folder_format
- group_index_format
- history_format
- index_format
- mailbox_folder_format
- mix_entry_format
- pattern_format
- pgp_entry_format
- query_format

**Comment by flatcap** [comment]
This has been superseded by #4202.

The PR contains everything except the arrow_cursor changes.
The PR works, is robust and the code has 100% documentation and 100% coverage :-)

**Comment by g0mb4** [comment]
Sorry for my absence, I'm a bit busy :( I'll try to keep up with the changes.

**Comment by flatcap** [comment]
no worries

**Pull request closed without merging**